File: lightbulb-accordionview/src/main/java/com/github/RooneyAndShadows/lightbulb/accordionview/AccordionView.java
package com.github.rooneyandshadows.lightbulb.accordionview;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.github.rooneyandshadows.java.commons.string.StringUtils;
import com.github.rooneyandshadows.lightbulb.accordionview.animation.AccordionAnimation;
import com.github.rooneyandshadows.lightbulb.accordionview.animation.AccordionShowHideAnimation;
import com.github.rooneyandshadows.lightbulb.accordionview.animation.AccordionTransitionAnimation;
import com.github.rooneyandshadows.lightbulb.commons.utils.DrawableUtils;
import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment;
import com.github.rooneyandshadows.lightbulb.dialogs.dialog_alert.AlertDialog;
import com.github.rooneyandshadows.lightbulb.dialogs.dialog_alert.AlertDialogBuilder;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageButton;
import androidx.appcompat.widget.LinearLayoutCompat;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.databinding.BindingAdapter;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

@SuppressWarnings("unused")
public class AccordionView extends LinearLayoutCompat {
    private LinearLayoutCompat accordionHeaderContainer;
    private ConstraintLayout accordionHeader;
    private RelativeLayout contentContainer;
    private TextView headingTextView;
    private AppCompatImageButton additionalInfoButton;
    private AppCompatImageButton expandButton;
    private int animationDuration = 250;
    private int backgroundCornerRadius = 0;
    private int headingTextSize;
    private int additionalInfoDrawableColor;
    private int expandDrawableColor;
    private int backgroundColor;
    private int headingTextColor;
    private int headingTextAppearance = -1;
    private Drawable expandIcon;
    private Drawable collapseIcon;
    private Drawable backgroundDrawable;
    private boolean expanded = false;
    private boolean expandable = true;
    private boolean expandOnHeadingClick = false;
    private String headingText;
    private ExpansionListeners expandListeners;
    private ExpansionListeners onGroupCheckedListener;
    private AccordionAnimation anim;
    private ContentPositionType contentPosition;
    private AccordionAnimationType animationType;
    private AlertDialog dialog;
    private boolean dialogEnabled;
    private String DIALOG_TAG;
    private String dialogTitle;
    private String dialogMessage;
    private String dialogButtonText;
    private boolean inflated;

    public AccordionView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        setSaveEnabled(true);
        readAttributes(context, attrs);
        inflate(getContext(), R.layout.view_accordion_layout, this);
        inflated = true;
        initView();
    }

    @Override
    public void setClipChildren(boolean clipChildren) {
        super.setClipChildren(clipChildren);
        if (inflated)
            setupClips();
    }

    @Override
    public void setClipToPadding(boolean clipToPadding) {
        super.setClipToPadding(clipToPadding);
        if (inflated)
            setupClips();
    }

    void setOnGroupCheckedListener(ExpansionListeners listener) {
        onGroupCheckedListener = listener;
    }

    public boolean isExpanded() {
        return expanded;
    }

    public void setExpandListeners(ExpansionListeners expandListeners) {
        this.expandListeners = expandListeners;
    }

    public void setDialogTitle(String dialogTitle) {
        this.dialogTitle = dialogTitle;
        dialog.setTitle(dialogTitle);
    }

    public void setDialogMessage(String dialogMessage) {
        this.dialogMessage = dialogMessage;
        dialog.setMessage(dialogMessage);
    }

    public void setDialogButtonText(String dialogButtonText) {
        this.dialogButtonText = dialogButtonText;
        initializeInformationButton();
    }

    public void setHeadingTextColor(int headingTextColor) {
        this.headingTextColor = headingTextColor;
        headingTextView.setTextColor(headingTextColor);
    }

    public void setHeadingTextSize(int headingTextSize) {
        this.headingTextSize = headingTextSize;
        headingTextView.setTextSize(headingTextSize);
    }

    public void setHeadingTextAppearance(int headingTextAppearance) {
        this.headingTextAppearance = headingTextAppearance;
        headingTextView.setTextAppearance(headingTextAppearance);
    }

    public void setAnimationDuration(int animationDuration) {
        this.animationDuration = animationDuration;
    }

    public void setBackgroundCornerRadius(int backgroundCornerRadius) {
        this.backgroundCornerRadius = backgroundCornerRadius;
        setupHeader();
    }

    public void setBackground(int backgroundColor) {
        this.backgroundColor = backgroundColor;
        setupHeader();
    }

    public void setBackground(Drawable backgroundDrawable) {
        this.backgroundDrawable = backgroundDrawable;
        setupHeader();
    }

    public void setExpandOnHeadingClick(boolean expandOnHeadingClick) {
        this.expandOnHeadingClick = expandOnHeadingClick;
        setupHeader();
    }

    public void setHeadingText(String headingText) {
        this.headingText = headingText;
        headingTextView.setText(headingText);
    }

    public void setExpandable(boolean expandable) {
        this.expandable = expandable;
        initializeExpandButton();
    }

    public void setContentPosition(ContentPositionType contentPosition) {
        ViewGroup oldContainer = contentContainer;
        this.contentPosition = contentPosition;
        RelativeLayout accordionBelowContent = findViewById(R.id.content_container_below);
        RelativeLayout accordionDefaultContent = findViewById(R.id.content_container_default);
        if (this.contentPosition == ContentPositionType.BELOW_HEADER)
            contentContainer = accordionBelowContent;
        else contentContainer = accordionDefaultContent;
        if (oldContainer != contentContainer)
            while (oldContainer.getChildCount() > 0) {
                contentContainer.addView(oldContainer.getChildAt(0));
                oldContainer.removeViewAt(0);
            }
        setupInitialExpandState();
    }

    public void expand(boolean animated) {
        AccordionView view = this;
        if (expandable && expanded || (animated && anim.hasRunningAnimation()))
            return;
        if (animated)
            anim.expand(animationDuration);
        else
            contentContainer.setVisibility(VISIBLE);
        expanded = true;
        expandButton.setImageDrawable(collapseIcon);
        if (expandListeners != null)
            expandListeners.execute(view, true);
        if (onGroupCheckedListener != null)
            onGroupCheckedListener.execute(view, true);
    }

    public void collapse(boolean animated) {
        AccordionView view = this;
        if (expandable && !expanded || (animated && anim.hasRunningAnimation()))
            return;
        if (animated)
            anim.collapse(animationDuration);
        else
            contentContainer.setVisibility(GONE);
        expanded = false;
        expandButton.setImageDrawable(expandIcon);
        if (expandListeners != null)
            onGroupCheckedListener.execute(view, false);
        if (onGroupCheckedListener != null)
            onGroupCheckedListener.execute(view, false);
    }

    @BindingAdapter("accordionHeadingText")
    public static void setHeadingText(AccordionView view, String text) {
        view.setHeadingText(text);
    }

    private void initView() {
        setOrientation(VERTICAL);
        selectChildren();
        setupClips();
        initializeHeader();
        initAnimation();
    }

    private void setupClips() {
        RelativeLayout accordionBelowContent = findViewById(R.id.content_container_below);
        RelativeLayout accordionDefaultContent = findViewById(R.id.content_container_default);
        accordionBelowContent.setClipToPadding(getClipToPadding());
        accordionDefaultContent.setClipToPadding(getClipToPadding());
        accordionHeaderContainer.setClipToPadding(getClipToPadding());
        accordionHeader.setClipToPadding(getClipToPadding());
        accordionBelowContent.setClipChildren(getClipChildren());
        accordionDefaultContent.setClipChildren(getClipChildren());
        accordionHeaderContainer.setClipChildren(getClipChildren());
        accordionHeader.setClipChildren(getClipChildren());
    }

    private void readAttributes(Context context, AttributeSet attrs) {
        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.AccordionView, 0, 0);
        try {
            DIALOG_TAG = a.getString(R.styleable.AccordionView_AV_DialogTag);
            dialogTitle = a.getString(R.styleable.AccordionView_AV_DialogTitle);
            dialogMessage = a.getString(R.styleable.AccordionView_AV_DialogMessage);
            headingText = a.getString(R.styleable.AccordionView_AV_HeadingText);
            dialogButtonText = a.getString(R.styleable.AccordionView_AV_DialogButtonText);
            if (StringUtils.isNullOrEmptyString(headingText))
                headingText = ResourceUtils.getPhrase(context, R.string.av_heading_default_text);
            if (StringUtils.isNullOrEmptyString(DIALOG_TAG))
                DIALOG_TAG = "ACCORDION_TOOLTIP_TAG";
            if (StringUtils.isNullOrEmptyString(dialogTitle))
                dialogTitle = ResourceUtils.getPhrase(context, R.string.av_default_dialog_title_text);
            if (StringUtils.isNullOrEmptyString(dialogMessage))
                dialogMessage = ResourceUtils.getPhrase(context, R.string.av_default_dialog_message_text);
            contentPosition = ContentPositionType.valueOf(a.getInt(R.styleable.AccordionView_AV_ContentPosition, 1));
            animationType = AccordionAnimationType.valueOf(a.getInt(R.styleable.AccordionView_AV_AnimationType, 1));
            additionalInfoDrawableColor = a.getColor(R.styleable.AccordionView_AV_InfoIconColor, ResourceUtils.getColorByAttribute(getContext(), R.attr.colorOnSurface));
            expandDrawableColor = a.getColor(R.styleable.AccordionView_AV_ExpandIconColor, ResourceUtils.getColorByAttribute(getContext(), R.attr.colorOnSurface));
            headingTextColor = a.getColor(R.styleable.AccordionView_AV_HeadingTextColor, -1);
            backgroundColor = a.getColor(R.styleable.AccordionView_AV_BackgroundColor, ResourceUtils.getColorByAttribute(getContext(), R.attr.colorSurface));
            headingTextSize = a.getDimensionPixelSize(R.styleable.AccordionView_AV_HeadingTextSize, -1);
            headingTextAppearance = a.getResourceId(R.styleable.AccordionView_AV_HeadingTextAppearance, R.style.Accordion_HeadingTextAppearance);
            backgroundDrawable = a.getDrawable(R.styleable.AccordionView_AV_BackgroundDrawable);
            animationDuration = a.getInteger(R.styleable.AccordionView_AV_AnimationDuration, animationDuration);
            backgroundCornerRadius = a.getInteger(R.styleable.AccordionView_AV_BackgroundCornerRadius, backgroundCornerRadius);
            expanded = a.getBoolean(R.styleable.AccordionView_AV_Expanded, expanded);
            expandable = a.getBoolean(R.styleable.AccordionView_AV_Expandable, expandable);
            expandOnHeadingClick = a.getBoolean(R.styleable.AccordionView_AV_ExpandOnHeadingClick, expandOnHeadingClick);
            dialogEnabled = a.getBoolean(R.styleable.AccordionView_AV_DialogEnabled, dialogEnabled);
        } finally {
            a.recycle();
        }
    }

    private void selectChildren() {
        accordionHeaderContainer = findViewById(R.id.accordion_header_container);
        accordionHeader = findViewById(R.id.accordion_header);
        RelativeLayout accordionBelowContent = findViewById(R.id.content_container_below);
        RelativeLayout accordionDefaultContent = findViewById(R.id.content_container_default);
        headingTextView = findViewById(R.id.accordionHeadingText);
        additionalInfoButton = findViewById(R.id.accordionInformationButton);
        expandButton = findViewById(R.id.accordionExpandButton);
        if (contentPosition == ContentPositionType.BELOW_HEADER)
            contentContainer = accordionBelowContent;
        else contentContainer = accordionDefaultContent;
    }

    private void initializeHeader() {
        headingTextView.setText(headingText);
        headingTextView.setTextAppearance(headingTextAppearance);
        if (headingTextColor != -1)
            headingTextView.setTextColor(headingTextColor);
        if (headingTextSize != -1)
            headingTextView.setTextSize(TypedValue.COMPLEX_UNIT_PX, headingTextSize);
        expandIcon = ResourceUtils.getDrawable(getContext(), R.drawable.icon_expand);
        collapseIcon = ResourceUtils.getDrawable(getContext(), R.drawable.icon_collapse);
        expandIcon.setTint(expandDrawableColor);
        collapseIcon.setTint(expandDrawableColor);
        setupHeader();
        setupInitialExpandState();
        if (isInEditMode())
            return;
        initializeInformationButton();
        initializeExpandButton();
        measureWithButtons();
    }

    private void measureWithButtons() {
        int matchParentMeasureSpec = View.MeasureSpec.makeMeasureSpec(((View) accordionHeader.getParent()).getWidth(), View.MeasureSpec.EXACTLY);
        int wrapContentMeasureSpec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED);
        expandButton.setVisibility(VISIBLE);
        additionalInfoButton.setVisibility(VISIBLE);
        accordionHeader.measure(matchParentMeasureSpec, wrapContentMeasureSpec);
        accordionHeader.setMinHeight(accordionHeader.getMeasuredHeight());
    }

    private void initializeInformationButton() {
        if (!dialogEnabled) {
            additionalInfoButton.setBackground(null);
            additionalInfoButton.setVisibility(GONE);
            return;
        }
        additionalInfoButton.setBackgroundResource(R.drawable.accordion_heading_button_background);
        FragmentManager manager = ((FragmentActivity) getContext()).getSupportFragmentManager();
        dialog = new AlertDialogBuilder(manager, DIALOG_TAG)
                .withDialogType(BaseDialogFragment.DialogTypes.BOTTOM_SHEET)
                .withTitle(dialogTitle)
                .withMessage(dialogMessage)
                .withCancelOnClickOutsude(true)
                .withPositiveButton(new BaseDialogFragment.DialogButtonConfiguration(dialogButtonText, true, true), null)
                .buildDialog();
        Drawable additionalInfoIcon = ResourceUtils.getDrawable(getContext(), R.drawable.icon_info);
        additionalInfoIcon.setTint(additionalInfoDrawableColor);
        additionalInfoButton.setImageDrawable(additionalInfoIcon);
        additionalInfoButton.setOnClickListener(view -> dialog.show());
    }

    private void setupHeader() {
        accordionHeaderContainer.setBackground(backgroundDrawable != null ? backgroundDrawable : DrawableUtils.getRoundedCornersDrawable(backgroundColor, backgroundCornerRadius));
        if (expandOnHeadingClick && expandable) {
            accordionHeader.setClickable(true);
            accordionHeader.setFocusable(true);
            accordionHeader.setOnClickListener(view -> {
                if (expanded) collapse(true);
                else expand(true);
            });
        } else {
            accordionHeader.setOnClickListener(null);
        }
    }

    private void setupInitialExpandState() {
        if (expanded)
            if (contentContainer.getVisibility() == GONE)
                contentContainer.setVisibility(VISIBLE);
        if (!expanded)
            if (contentContainer.getVisibility() == VISIBLE)
                contentContainer.setVisibility(GONE);
    }

    private void initializeExpandButton() {
        if (expanded) {
            if (expandable) expandButton.setImageDrawable(collapseIcon);
        } else {
            if (expandable) expandButton.setImageDrawable(expandIcon);
        }
        if (expandable) {
            expandButton.setBackgroundResource(R.drawable.accordion_heading_button_background);
            expandButton.setVisibility(VISIBLE);
            expandButton.setOnClickListener(view -> {
                if (expanded) collapse(true);
                else expand(true);
            });
        } else {
            expandButton.setBackground(null);
            expandButton.setVisibility(GONE);
        }
    }

    private void initAnimation() {
        switch (animationType) {
            case ANIM_NONE:
                anim = new AccordionShowHideAnimation(contentContainer);
                break;
            case ANIM_HEIGHT_TRANSITION:
                anim = new AccordionTransitionAnimation(contentContainer);
                break;
        }
    }

    @Override
    public void addView(View child, int index, ViewGroup.LayoutParams params) {
        if (contentContainer != null) { //check if is initial inflation
            contentContainer.addView(child, index, params);
        } else
            super.addView(child, index, params);
    }

    @Override
    protected void dispatchSaveInstanceState(SparseArray<Parcelable> container) {
        dispatchFreezeSelfOnly(container);
    }

    @Override
    protected void dispatchRestoreInstanceState(SparseArray<Parcelable> container) {
        dispatchThawSelfOnly(container);
    }

    @Override
    public Parcelable onSaveInstanceState() {
        Parcelable superState = super.onSaveInstanceState();
        SavedState myState = new SavedState(superState);
        myState.title = this.dialogTitle;
        myState.subtitle = this.dialogMessage;
        myState.buttonText = this.dialogButtonText;
        myState.expanded = this.expanded;
        myState.backgroundColor = this.backgroundColor;
        myState.cornerRadius = this.backgroundCornerRadius;
        myState.headingTextSize = this.headingTextSize;
        myState.headingTextColor = this.headingTextColor;
        myState.headingTextAppearance = this.headingTextAppearance;
        myState.animationDuration = this.animationDuration;
        myState.visibility = this.getVisibility();
        return myState;
    }

    @Override
    public void onRestoreInstanceState(Parcelable state) {
        SavedState savedState = (SavedState) state;
        super.onRestoreInstanceState(savedState.getSuperState());
        this.dialogTitle = savedState.title;
        this.dialogMessage = savedState.subtitle;
        this.dialogButtonText = savedState.buttonText;
        this.expanded = savedState.expanded;
        this.backgroundColor = savedState.backgroundColor;
        this.backgroundCornerRadius = savedState.cornerRadius;
        this.headingTextSize = savedState.headingTextSize;
        this.headingTextColor = savedState.headingTextColor;
        this.headingTextAppearance = savedState.headingTextAppearance;
        this.animationDuration = savedState.animationDuration;
        this.setVisibility(savedState.visibility);
        initView();
    }

    private static class SavedState extends BaseSavedState {
        private String title;
        private String subtitle;
        private String buttonText;
        private boolean expanded;
        private int cornerRadius;
        private int headingTextSize;
        private int headingTextColor;
        private int headingTextAppearance;
        private int backgroundColor;
        private int animationDuration;
        private int visibility;

        SavedState(Parcelable superState) {
            super(superState);
        }

        private SavedState(Parcel in) {
            super(in);
            title = in.readString();
            subtitle = in.readString();
            buttonText = in.readString();
            cornerRadius = in.readInt();
            headingTextSize = in.readInt();
            headingTextColor = in.readInt();
            headingTextAppearance = in.readInt();
            backgroundColor = in.readInt();
            animationDuration = in.readInt();
            visibility = in.readInt();
            expanded = in.readInt() == 1;
        }

        @Override
        public void writeToParcel(Parcel out, int flags) {
            super.writeToParcel(out, flags);
            out.writeString(title);
            out.writeString(subtitle);
            out.writeString(buttonText);
            out.writeInt(cornerRadius);
            out.writeInt(headingTextSize);
            out.writeInt(headingTextColor);
            out.writeInt(headingTextAppearance);
            out.writeInt(backgroundColor);
            out.writeInt(animationDuration);
            out.writeInt(visibility);
            out.writeInt(expanded ? 1 : 0);
        }

        public static final Creator<SavedState> CREATOR
                = new Creator<SavedState>() {
            public SavedState createFromParcel(Parcel in) {
                return new SavedState(in);
            }

            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };
    }

    public interface ExpansionListeners {
        public void execute(AccordionView view, boolean expanded);
    }

    public enum ContentPositionType {
        INSIDE_HEADER(1),
        BELOW_HEADER(2);

        private final int value;
        private static final SparseArray<ContentPositionType> map = new SparseArray<>();

        ContentPositionType(int value) {
            this.value = value;
        }

        static {
            for (ContentPositionType contentPosition : ContentPositionType.values()) {
                map.put(contentPosition.value, contentPosition);
            }
        }

        public static ContentPositionType valueOf(int contentPosition) {
            return map.get(contentPosition);
        }

        public int getValue() {
            return value;
        }
    }

    public enum AccordionAnimationType {
        ANIM_NONE(1),
        ANIM_HEIGHT_TRANSITION(2);

        private final int value;
        private static final SparseArray<AccordionAnimationType> map = new SparseArray<>();

        AccordionAnimationType(int value) {
            this.value = value;
        }

        static {
            for (AccordionAnimationType animation : AccordionAnimationType.values()) {
                map.put(animation.value, animation);
            }
        }

        public static AccordionAnimationType valueOf(int animation) {
            return map.get(animation);
        }

        public int getValue() {
            return value;
        }
    }
}

File: lightbulb-accordionview/src/main/java/com/github/RooneyAndShadows/lightbulb/accordionview/AccordionView.kt
package com.github.rooneyandshadows.lightbulb.accordionview

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.os.Parcel
import android.os.Parcelable
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageButton
import androidx.appcompat.widget.LinearLayoutCompat
import androidx.databinding.BindingAdapter
import com.github.rooneyandshadows.java.commons.string.StringUtils
import com.github.rooneyandshadows.lightbulb.accordionview.AccordionView.AccordionAnimationType.*
import com.github.rooneyandshadows.lightbulb.accordionview.AccordionView.AccordionBackground.*
import com.github.rooneyandshadows.lightbulb.accordionview.animation.AccordionAnimation
import com.github.rooneyandshadows.lightbulb.accordionview.animation.AccordionShowHideAnimation
import com.github.rooneyandshadows.lightbulb.accordionview.animation.AccordionTransitionAnimation
import com.github.rooneyandshadows.lightbulb.commons.utils.ParcelUtils
import com.github.rooneyandshadows.lightbulb.commons.utils.ResourceUtils

@Suppress("MemberVisibilityCanBePrivate", "unused", "LeakingThis")
open class AccordionView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = R.attr.accordionViewStyle,
    defStyleRes: Int = R.style.AccordionViewDefaultStyle,
) : RelativeLayout(context, attrs, defStyleAttr, defStyleRes) {
    private val stateExpanded = intArrayOf(R.attr.av_state_expanded)
    private val stateCollapsed = intArrayOf(-R.attr.av_state_expanded)
    private val accordionHeader: LinearLayoutCompat by lazy {
        return@lazy findViewById(R.id.accordion_header)!!
    }
    private val contentContainer: RelativeLayout by lazy {
        return@lazy findViewById(R.id.accordion_content)!!
    }
    private val headingTextView: TextView by lazy {
        return@lazy findViewById(R.id.accordion_heading_text)!!
    }
    private val expandButton: AppCompatImageButton by lazy {
        return@lazy findViewById(R.id.accordion_expand_button)
    }
    private var expandListener: OnExpandedChangeListener? = null
    private var onGroupCheckedListener: OnExpandedChangeListener? = null
    private var anim: AccordionAnimation? = null
    private var inflated = false
    private var expandIcon: Drawable? = null
        set(value) {
            field = value
            field?.setTint(expandDrawableColor)
        }
    var animationDuration = 250
    var expandable = true
        set(value) {
            field = value
            syncExpandButton()
        }
    var headingTextSize: Int = -1
        set(value) {
            field = value
            if (value == -1) return
            headingTextView.textSize = headingTextSize.toFloat()
        }
    var animationType: AccordionAnimationType = ANIM_HEIGHT_TRANSITION
        set(value) {
            if (field == value) return
            field = value
            syncExpandButton()
        }
    var expandDrawableColor: Int = Color.TRANSPARENT
        set(value) {
            field = value
            expandIcon?.setTint(field)
        }
    var headingTextColor = -1
        set(value) {
            field = value
            if (value == -1) return
            headingTextView.setTextColor(headingTextColor)
        }
    var headingTextAppearance = -1
        set(value) {
            field = value
            if (value == -1) return
            headingTextView.setTextAppearance(headingTextAppearance)
        }
    var expandOnHeadingClick = true
        set(value) {
            if (field == value) return
            field = value
            syncExpandButton()
        }
    var headingText: String? = null
        set(value) {
            field = value
            headingTextView.text = headingText
        }
    var accordionBackground: AccordionBackground = BG_CARD
        set(value) {
            field = value
            elevation = 0F
            background = when (value) {
                BG_CARD -> {
                    elevation = ResourceUtils.getDimenById(context, R.dimen.av_bg_card_elevation)
                    ResourceUtils.getDrawable(context, R.drawable.av_bg_card)
                }
                BG_STROKED -> {
                    ResourceUtils.getDrawable(context, R.drawable.av_bg_stroke)
                }
                else -> null
            }
        }
    var isExpanded = false
        private set(value) {
            if (field == value) return
            field = value
            refreshDrawableState()
        }

    init {
        isSaveEnabled = true
        inflate(getContext(), R.layout.view_accordion_layout, this)
        readAttributes(context, attrs, defStyleAttr, defStyleRes)
        setupClips()
        initializeHeader()
        syncAnimation()
        inflated = true
    }

    @Override
    override fun addView(child: View, index: Int, params: ViewGroup.LayoutParams) {
        if (!inflated) super.addView(child, index, params)
        else contentContainer.addView(child, index, params)
    }

    @Override
    override fun setClipChildren(clipChildren: Boolean) {
        super.setClipChildren(clipChildren)
        if (!inflated) return
        setupClips()
    }

    @Override
    override fun setClipToPadding(clipToPadding: Boolean) {
        super.setClipToPadding(clipToPadding)
        if (!inflated) return
        setupClips()
    }

    @Override
    override fun onCreateDrawableState(extraSpace: Int): IntArray? {
        val drawableState = super.onCreateDrawableState(extraSpace + 1)
        if (isExpanded) {
            mergeDrawableStates(drawableState, stateExpanded)
        }
        return drawableState
    }

    fun setOnGroupCheckedListener(listener: OnExpandedChangeListener?) {
        onGroupCheckedListener = listener
    }

    fun setOnExpandListeners(expandListener: OnExpandedChangeListener?) {
        this.expandListener = expandListener
    }

    fun expand(animated: Boolean) {
        if (expandable && isExpanded) return
        if (animated) anim!!.expand(animationDuration) else contentContainer.visibility = VISIBLE
        isExpanded = true
        expandButton.animate().setDuration(animationDuration.toLong()).rotation(180f).start()
        if (expandListener != null) expandListener!!.execute(this, true)
        if (onGroupCheckedListener != null) onGroupCheckedListener!!.execute(this, true)
    }

    fun collapse(animated: Boolean) {
        if (expandable && !isExpanded) return
        if (animated) anim!!.collapse(animationDuration) else contentContainer.visibility = GONE
        isExpanded = false
        expandButton.animate().setDuration(animationDuration.toLong()).rotation(0f).start()
        if (expandListener != null) onGroupCheckedListener!!.execute(this, false)
        if (onGroupCheckedListener != null) onGroupCheckedListener!!.execute(this, false)
    }


    private fun setupClips() {
        contentContainer.clipToPadding = clipToPadding
        accordionHeader.clipToPadding = clipToPadding
        contentContainer.clipChildren = clipChildren
        accordionHeader.clipChildren = clipChildren
    }

    private fun readAttributes(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) {
        val attr = context.theme.obtainStyledAttributes(
            attrs,
            R.styleable.AccordionView,
            defStyleAttr,
            defStyleRes
        )
        attr.apply {
            try {
                headingText = getString(R.styleable.AccordionView_av_heading_text)
                if (StringUtils.isNullOrEmptyString(headingText)) headingText =
                    ResourceUtils.getPhrase(context, R.string.av_heading_default_text)
                animationType =
                    AccordionAnimationType.valueOf(getInt(R.styleable.AccordionView_av_animation, animationType.value))
                accordionBackground = AccordionBackground.valueOf(getInt(R.styleable.AccordionView_av_background, 1))
                expandDrawableColor = getColor(
                    R.styleable.AccordionView_av_expand_icon_color,
                    ResourceUtils.getColorByAttribute(getContext(), R.attr.colorOnSurface)
                )
                headingTextColor = getColor(R.styleable.AccordionView_av_heading_text_color, -1)
                headingTextSize = getDimensionPixelSize(R.styleable.AccordionView_av_heading_text_Size, -1)
                headingTextAppearance = getResourceId(
                    R.styleable.AccordionView_av_heading_text_appearance,
                    R.style.AccordionHeadingTextAppearance
                )
                animationDuration = getInteger(R.styleable.AccordionView_av_animation_duration, animationDuration)
                isExpanded = getBoolean(R.styleable.AccordionView_av_expanded, isExpanded)
                expandable = getBoolean(R.styleable.AccordionView_av_expandable, expandable)
                expandOnHeadingClick = getBoolean(R.styleable.AccordionView_av_expand_on_heading_click, expandOnHeadingClick)

            } finally {
                recycle()
            }
        }
    }

    private fun syncExpandButton() {
        if (expandOnHeadingClick && expandable) {
            accordionHeader.isClickable = true
            accordionHeader.isFocusable = true
            accordionHeader.setOnClickListener {
                if (isExpanded) collapse(true)
                else expand(true)
            }
        } else {
            accordionHeader.setOnClickListener(null)
        }
    }

    private fun initializeHeader() {
        headingTextView.text = headingText
        if (headingTextColor != -1) headingTextView.setTextColor(headingTextColor)
        if (headingTextSize != -1) headingTextView.setTextSize(TypedValue.COMPLEX_UNIT_PX, headingTextSize.toFloat())
        setupInitialExpandState()
        initializeExpandButton()
        syncExpandButton()
    }

    private fun setupInitialExpandState() {
        if (isExpanded) {
            expandButton.rotation = 180F
            if (contentContainer.visibility == GONE) contentContainer.visibility = VISIBLE

        }
        if (!isExpanded) {
            expandButton.rotation = 0F
            if (contentContainer.visibility == VISIBLE) contentContainer.visibility = GONE
        }
    }

    private fun initializeExpandButton() {
        expandIcon = ResourceUtils.getDrawable(context, R.drawable.av_icon_expand)
        expandIcon!!.setTint(expandDrawableColor)
        expandButton.apply {
            if (expandable) {
                expandButton.setImageDrawable(expandIcon)
                expandButton.setBackgroundResource(R.drawable.av_heading_button_background)
                expandButton.setOnClickListener { if (isExpanded) collapse(true) else expand(true) }
                expandButton.visibility = VISIBLE
                return@apply
            }
            expandButton.background = null
            expandButton.visibility = GONE
        }
    }

    private fun syncAnimation() {
        anim = when (animationType) {
            ANIM_NONE -> AccordionShowHideAnimation(contentContainer)
            ANIM_HEIGHT_TRANSITION -> AccordionTransitionAnimation(contentContainer)
        }
    }

    //@Override
    //protected void dispatchSaveInstanceState(SparseArray<Parcelable> container) {
    //    dispatchFreezeSelfOnly(container);
    //}

    //@Override
    //protected void dispatchRestoreInstanceState(SparseArray<Parcelable> container) {
    //    dispatchThawSelfOnly(container);
    //}

    public override fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState()
        val myState = SavedState(superState)
        myState.expanded = isExpanded
        myState.headingTextSize = headingTextSize
        myState.headingTextColor = headingTextColor
        myState.headingTextAppearance = headingTextAppearance
        myState.animationDuration = animationDuration
        myState.visibility = this.visibility
        initializeHeader()
        return myState
    }

    public override fun onRestoreInstanceState(state: Parcelable) {
        val savedState = state as SavedState
        super.onRestoreInstanceState(savedState.superState)
        isExpanded = savedState.expanded
        headingTextSize = savedState.headingTextSize
        headingTextColor = savedState.headingTextColor
        headingTextAppearance = savedState.headingTextAppearance
        animationDuration = savedState.animationDuration
        this.visibility = savedState.visibility
        initializeHeader()
    }

    fun interface OnExpandedChangeListener {
        fun execute(view: AccordionView, expanded: Boolean)
    }

    private class SavedState : BaseSavedState {
        var expanded = false
        var cornerRadius = 0
        var headingTextSize = 0
        var headingTextColor = 0
        var headingTextAppearance = 0
        var backgroundColor = 0
        var animationDuration = 0
        var visibility = 0
        var accordionBackground: AccordionBackground = BG_CARD

        constructor(superState: Parcelable?) : super(superState)

        private constructor(parcel: Parcel) : super(parcel) {
            ParcelUtils.apply {
                cornerRadius = readInt(parcel)!!
                headingTextSize = readInt(parcel)!!
                headingTextColor = readInt(parcel)!!
                headingTextAppearance = readInt(parcel)!!
                backgroundColor = readInt(parcel)!!
                animationDuration = readInt(parcel)!!
                visibility = readInt(parcel)!!
                expanded = readBoolean(parcel)!!
                accordionBackground = AccordionBackground.valueOf(readInt(parcel)!!)
            }
        }

        override fun writeToParcel(out: Parcel, flags: Int) {
            super.writeToParcel(out, flags)
            ParcelUtils.apply {
                writeInt(out, cornerRadius)
                writeInt(out, headingTextSize)
                writeInt(out, headingTextColor)
                writeInt(out, headingTextAppearance)
                writeInt(out, backgroundColor)
                writeInt(out, animationDuration)
                writeInt(out, visibility)
                writeBoolean(out, expanded)
                writeInt(out, accordionBackground.value)
            }
        }

        override fun describeContents(): Int {
            return 0
        }

        companion object CREATOR : Parcelable.Creator<SavedState> {
            override fun createFromParcel(parcel: Parcel): SavedState {
                return SavedState(parcel)
            }

            override fun newArray(size: Int): Array<SavedState?> {
                return arrayOfNulls(size)
            }
        }
    }

    enum class AccordionAnimationType(val value: Int) {
        ANIM_NONE(1),
        ANIM_HEIGHT_TRANSITION(2);

        companion object {
            fun valueOf(value: Int) = values().first { it.value == value }
        }
    }

    enum class AccordionBackground(val value: Int) {
        BG_CARD(1),
        BG_STROKED(2),
        BG_CUSTOM(3);

        companion object {
            fun valueOf(value: Int) = values().first { it.value == value }
        }
    }

    companion object {
        @BindingAdapter("accordionHeadingText")
        fun setHeadingText(view: AccordionView, text: String?) {
            view.headingText = text
        }
    }
}